### PR TITLE
Dev/delivry 17805/unit tests

### PR DESCRIPTION
**Pull request by morisgateno-appsflyer**

### CODE REVIEW

File: app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegrationTests.java [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+    }
+
+//    @Test
+//    public void testAppsflyerIntegration_ctor_nullFlow() throws Exception {
+//        Context context = null;
+//        Logger logger = null;
+//        AppsFlyerLib appsflyer = null;
+//        String appsflyerDevKey = null;
+////        the line below is a problem that needs to be sorted.
+////        boolean isDebug = logger.logLevel != Analytics.LogLevel.NONE;
```
[the remark this answers is not in the document]
Comment (morisgateno-appsflyer) [reply app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegrationTests.java:66]: I was referring to the logger.logLevel that could be null, but you are right there isn't any problem here.

File: app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegrationTests.java [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+        Assert.assertTrue(appsflyerIntegration.appsflyer == appsflyer);
+        Assert.assertTrue(appsflyerIntegration.logger == logger);
+        Field field = AppsflyerIntegration.class.getDeclaredField("context");
+        field.setAccessible(true);
+
+        Context contextInappsflyerIntegration = (Context) field.get(appsflyerIntegration);
+
+        Assert.assertTrue(contextInappsflyerIntegration == context);
+//        checking the static clause
+        Assert.assertTrue(appsflyerIntegration.MAPPER.get("revenue")== AFInAppEventParameterName.REVENUE);
```
[the remark this answers is not in the document]
Comment (morisgateno-appsflyer) [reply app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegrationTests.java]: absolutely right. And by making AssertEquals, it invokes the .equals() of each object? 
if it doesn't have equals implements, it compares pointers?

[the remark this answers is not in the document]
Comment (pazlavi) [reply app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegrationTests.java]: Yes, `AssertEquals` is using the `equals` method defined in [`Object`](https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html#equals(java.lang.Object)) class. 
In Java, every class is inherited from Object and can override the implementation of this method.

File: app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegrationTests.java [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+        Assert.assertEquals(appsflyerIntegration.logger, logger);
+        Field field = AppsflyerIntegration.class.getDeclaredField("context");
+        field.setAccessible(true);
+
+        Context contextInappsflyerIntegration = (Context) field.get(appsflyerIntegration);
+
+        Assert.assertEquals(contextInappsflyerIntegration, context);
+//        checking the static clause
+        Assert.assertEquals(appsflyerIntegration.MAPPER.get("revenue"), AFInAppEventParameterName.REVENUE);
+        Assert.assertEquals(appsflyerIntegration.MAPPER.get("currency"), AFInAppEventParameterName.CURRENCY);
```
Comment (pazlavi) [inline app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegrationTests.java:50]: You should access static members using the class name and not via an instance (`AppsflyerIntegration`)

File: app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegrationTests.java [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+
+        reset(context,appsflyer);
+    }
+
+    @Test
+    public void testAppsflyerIntegration_setManualMode_happyFlow() throws Exception {
```
Comment (pazlavi) [inline app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegrationTests.java:56]: You don't need to declare  `throws Exception` on every test. 
Tests are throwing `AssertionError`.
An Error is a subclass of Throwable that indicates serious problems that a reasonable application should not try to catch.  In our case, this is an indication that our assertion failed and we don't need to handle this on runtime

File: app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegrationTests.java [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+        Assert.assertTrue(AppsflyerIntegration.manualMode);
+        AppsflyerIntegration.setManualMode(false);
+        Assert.assertFalse(AppsflyerIntegration.manualMode);
+    }
+
+    @Test
+    public void testAppsflyerIntegration_startAppsFlyer_happyFlow() throws Exception {
+        MockedStatic<AppsFlyerLib> staticAppsFlyerLib = mockStatic(AppsFlyerLib.class);
+        AppsFlyerLib appsFlyerLib = mock(AppsFlyerLib.class);
+        staticAppsFlyerLib.when(AppsFlyerLib::getInstance).thenReturn(appsFlyerLib);
```
Comment (pazlavi) [inline app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegrationTests.java:68]: Duplicate code, you should avoid writing the same code in multiple tests.   
You can extract it to a new helper method like `mockAppsFlyerLib()`

File: app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegrationTests.java [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+        AppsflyerIntegration.deepLinkListener = mock(AppsflyerIntegration.ExternalDeepLinkListener.class);
+
+        Integration<AppsFlyerLib> integration= (Integration<AppsFlyerLib>) AppsflyerIntegration.FACTORY.create(settings,analytics);
+
+        verify(appsFlyerLib).setDebugLog(logger.logLevel!=Analytics.LogLevel.NONE);
+        ArgumentCaptor<AppsflyerIntegration.ConversionListener> captorListener = ArgumentCaptor.forClass(AppsflyerIntegration.ConversionListener.class);
+        ArgumentCaptor<String> captorDevKey = ArgumentCaptor.forClass(String.class);
+        ArgumentCaptor<Context> captorContext = ArgumentCaptor.forClass(Context.class);
+        verify(appsFlyerLib).init(captorDevKey.capture(), captorListener.capture() , captorContext.capture());
+        Assert.assertNotEquals(captorListener.getValue(), null);
```
Comment (pazlavi) [inline app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegrationTests.java:115]: You can use `assertNotNull`

File: app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegrationTests.java [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+        Assert.assertTrue(captorListener.getValue() instanceof AppsflyerIntegration.ConversionListener);
```
Comment (pazlavi) [inline app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegrationTests.java:116]: Condition `captorListener.getValue() instanceof AppsflyerIntegration.ConversionListener` is redundant and can be replaced with a null check which is what you did in the previous line.

File: app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegrationTests.java (hunk 4 above)
Comment (morisgateno-appsflyer) [reply app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegrationTests.java:56]: Ahh OK.

Comment (morisgateno-appsflyer) [reply app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegrationTests.java:56]: But I see for some places - where I use reflection I need to make a `try catch` or `throws Exception`.
What should I prefer?

File: app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegrationTests.java [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+        Assert.assertEquals(appsflyerIntegration.getUnderlyingInstance(),appsFlyerLib);
+
+        reset(appsFlyerLib);
+    }
+
+    @Test
+    public void testAppsflyerIntegration_identify_happyFlow() throws Exception {
+        AppsFlyerLib appsFlyerLib = mock(AppsFlyerLib.class);
+        Logger logger = spy(new Logger("test", Analytics.LogLevel.INFO));
+        AppsflyerIntegration appsflyerIntegration = spy(new AppsflyerIntegration(null,logger,appsFlyerLib,null));
```
Comment (pazlavi) [inline app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegrationTests.java:186]: `spy` is redundant

File: app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegrationTests.java [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+        appsflyerIntegration.identify(identifyPayload);
+
+        verify(logger, never()).verbose(any());
+        Field customerUserIdField = AppsflyerIntegration.class.getDeclaredField("customerUserId");
+        customerUserIdField.setAccessible(true);
+        String customerUserIdInappsflyerIntegration = (String) customerUserIdField.get(appsflyerIntegration);
+        Assert.assertEquals(customerUserIdInappsflyerIntegration, "moris");
+        Field currencyCodeField = AppsflyerIntegration.class.getDeclaredField("currencyCode");
+        currencyCodeField.setAccessible(true);
+        String currencyCodeInappsflyerIntegration = (String) currencyCodeField.get(appsflyerIntegration);
```
Comment (pazlavi) [inline app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegrationTests.java:203]: you can create a helper class to get private fields and avoid code duplications

File: app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegrationTests.java [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+        Assert.assertEquals(currencyCodeInappsflyerIntegration, "ILS");
+
+        reset(appsFlyerLib,identifyPayload,traits);
+    }
+
+    @Test
+    public void testAppsflyerIntegration_identify_nilflow() throws Exception {
+        Logger logger = spy(new Logger("test", Analytics.LogLevel.INFO));
+        AppsflyerIntegration appsflyerIntegration = spy(new AppsflyerIntegration(null,logger,null,null));
```
Comment (pazlavi) [inline app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegrationTests.java:212]: spy is redundant

File: app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegrationTests.java (hunk 8 above)
Comment (morisgateno-appsflyer) [reply app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegrationTests.java:187]: The reason I use it is that in the function - identity - the logger calls its function 'verbose' and I want to stub it so I can't just mock it, I must partially mock it.
Maybe I'm wrong..

File: app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegration_ConversionListenerTests.java [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+import com.segment.analytics.Analytics;
+import com.segment.analytics.Properties;
+import com.segment.analytics.ValueMap;
+import static  org.mockito.Mockito.*;
+import java.lang.reflect.Method;
+import java.util.HashMap;
+import java.util.Map;
+
+@RunWith(AndroidJUnit4.class)
+public class AppsflyerIntegration_ConversionListenerTests {
```
Comment (pazlavi) [inline app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegration_ConversionListenerTests.java:20]: Please change the class name to match the regex   `[A-Z][A-Za-z\d]*Test(s|Case)?|Test[A-Z][A-Za-z\d]*`

File: app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegration_ConversionListenerTests.java [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+
+    @Test
+    public void testAppsflyerIntegration_ConversionListener_ctor_happyFlow() {
+        Analytics analytics = mock(Analytics.class);
+        AppsflyerIntegration.ConversionListener conversionListener = spy(new AppsflyerIntegration.ConversionListener(analytics));
```
Comment (pazlavi) [inline app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegration_ConversionListenerTests.java:25]: redundant spy

File: app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegration_ConversionListenerTests.java [hunk 13]
```diff
@@ -0,0 +1,274 @@
...
+
+        Assert.assertEquals(conversionListener.analytics, analytics);
+
+        reset(analytics,conversionListener);
+    }
+
+    @Test
+    public void testAppsflyerIntegration_ConversionListener_ctor_nullFlow() {
+        Analytics analytics = null;
+        AppsflyerIntegration.ConversionListener conversionListener = spy(new AppsflyerIntegration.ConversionListener(analytics));
```
Comment (pazlavi) [inline app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegration_ConversionListenerTests.java:35]: redundant spy

File: app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegration_ConversionListenerTests.java [hunk 14]
```diff
@@ -0,0 +1,274 @@
...
+        Analytics analytics = mock(Analytics.class);
+        Map<String, Object> conversionData =  new ValueMap();
+        Application app = mock(Application.class);
+        Context context = mock(Context.class);
+        SharedPreferences sharedPreferences = mock(SharedPreferences.class);
+        when(analytics.getApplication()).thenReturn(app);
+        when(app.getApplicationContext()).thenReturn(context);
+        when(context.getSharedPreferences("appsflyer-segment-data",0)).thenReturn(sharedPreferences);
+        when(sharedPreferences.getBoolean("AF_onConversion_Data",false)).thenReturn(true);
+        AppsflyerIntegration.ConversionListener conversionListener = spy(new AppsflyerIntegration.ConversionListener(analytics));
```
Comment (pazlavi) [inline app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegration_ConversionListenerTests.java:55]: redundant spy

File: app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegration_ConversionListenerTests.java [hunk 15]
```diff
@@ -0,0 +1,274 @@
...
+        conversionListener.onAttributionFailure(errorMsg);
+        verify(AppsflyerIntegration.conversionListener,times(1)).onAttributionFailure(null);
+
+        reset(analytics,conversionListener,AppsflyerIntegration.conversionListener);
+    }
+
+    @Test
+    public void testAppsflyerIntegration_ConversionListener_trackInstallAttributed_happyFlow() {
+        Analytics analytics =mock(Analytics.class);
+        Map<String, Object> attributionData = new HashMap<String, Object>()
```
Comment (pazlavi) [inline app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegration_ConversionListenerTests.java:93]: It's better to write like this: 
```java 
        Map<String, Object> attributionData = new HashMap<>();
        attributionData.put("media_source", "media_source_moris");
        attributionData.put("campaign", "campaign_moris");
        attributionData.put("adgroup", "adgroup_moris");
```

File: app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegration_ConversionListenerTests.java [hunk 16]
```diff
@@ -0,0 +1,274 @@
...
+        setFlagMethod.setAccessible(true);
+
+        setFlagMethod.invoke(conversionListener,key,value);
+
+        if(android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.GINGERBREAD){
+            verify(editor,times(1)).apply();
+        }
+        else{
+            verify(editor,times(1)).commit();
+        }
```
Comment (pazlavi) [inline app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegration_ConversionListenerTests.java:208]: The first condition is always true in our case.

Comment (morisgateno-appsflyer) [reply app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegration_ConversionListenerTests.java:208]: This code is from the AppsflyerIntegration class(I just want to check the correct method is called in this case), So if it is redundant here, is it redundant there?

Comment (pazlavi) [reply app/src/test/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegration_ConversionListenerTests.java:208]: Yes, our min OS version is way above GINGERBREAD

**Pull request merged**